compiler/rustc_resolve/src/imports.rs

```diff
@@ -554,6 +555,40 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         }
     }
 
+    pub(crate) fn check_local_binding_shadows_glob_reexport(&mut self) {
+        for module in self.arenas.local_modules().iter() {
+            let resolutions = self.resolutions(module).borrow();
+            for (_, resolution) in resolutions.iter() {
+                let resolution = resolution.borrow();
+                if let Some(binding) = resolution.binding
+                && let Some(glob_binding) = resolution.shadowed_glob
+                && let NameBinding { kind: NameBindingKind::Import { import, .. }, .. } = glob_binding
+                && let Some(import_id) = import.id()
+                && let import_local_def_id = self.local_def_id(import_id)
+                && self.effective_visibilities.is_exported(import_local_def_id)
+                && glob_binding.vis.is_public()
+                && !binding.vis.is_public()
+                && let Some(def_id) = binding.res().opt_def_id()
+                && let Some(local_def_id) = def_id.as_local()
+                && let Some(ns) = binding.res().ns()
+                {
+                    self.lint_buffer.buffer_lint_with_diagnostic(
+                        LOCAL_BINDING_SHADOWS_GLOB_REEXPORT,
+                        self.def_id_to_node_id[local_def_id],
```

Comment:
This is not a right node id to use here, you need an id of `resolution.binding` itself, not its target (i.e. id of the import item `use foo::bar;`, not id of `bar`).